File: backend/app/schemas/dev.py
"""Pydantic schemas for the stateless /dev/simulate endpoint."""
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, Field

class SimPrice(BaseModel):
    valid_from: datetime
    price_pence: float

class SimImmersionRule(BaseModel):
    rule_name: str = "rule"
    priority: int = 10
    action: str = "ON"
    logic_operator: str = "AND"
    is_enabled: bool = True
    price_enabled: bool = False
    price_operator: str = "<"
    price_threshold_pence: float = 15.0
    soc_enabled: bool = False
    soc_operator: str = ">"
    soc_threshold_percent: float = 50.0
    solar_enabled: bool = False
    solar_operator: str = ">"
    solar_threshold_kw: float = 2.0
    temp_enabled: bool = False
    temp_operator: str = "<"
    temp_threshold_c: float = 50.0
    time_enabled: bool = False
    time_start: str = "00:00"
    time_end: str = "23:59"

class SimImmersionInput(BaseModel):
    battery_soc: float = 60.0
    solar_power_kw: float = 3.0
    current_price_pence: float = 15.0
    current_temp_c: Optional[float] = None
    rules: List[SimImmersionRule] = []

class SimRequest(BaseModel):
    battery_soc: float = Field(50.0, ge=0, le=100)
    solar_power_kw: float = Field(10.6, ge=0)
    solar_profile: str = "sunny"
    solar_scale: float = Field(1.0, ge=0.0, le=1.0)
    live_charge_rate_kw: Optional[float] = None
    live_battery_voltage_v: Optional[float] = None
    battery_capacity_kwh: float = 20.0
    battery_max_charge_kw: float = 10.5
    battery_max_discharge_kw: float = 5.0
    battery_efficiency: float = 0.95
    battery_min_soc: int = 10
    battery_max_soc: int = 100
    battery_voltage_v: float = 48.0
    grid_import_limit_kw: float = 15.0
    grid_export_limit_kw: float = 5.0
    export_price_pence: float = 15.0
    assumed_load_kw: float = 2.0
    force_charge_threshold_kw: float = 0.5
    force_discharge_threshold_kw: float = 0.5
    force_discharge_export_min_kw: float = 0.05
    optimization_horizon_hours: int = 24
    prices: List[SimPrice] = []
    immersion: Optional[SimImmersionInput] = None

class SimPeriodResult(BaseModel):
    slot: int
    valid_from: str
    price_pence: float
    solar_kw: float
    charge_kw: float
    discharge_kw: float
    grid_import_kw: float
    grid_export_kw: float
    soc_kwh: float
    soc_pct: float

class SimImmersionResult(BaseModel):
    action: bool
    source: str
    reason: str

class SimResponse(BaseModel):
    recommended_mode: str
    decision_reason: str
    optimization_status: str
    objective_value: Optional[float]
    optimization_time_ms: float
    recommended_discharge_current: int
    periods: List[SimPeriodResult]
    immersion: Optional[SimImmersionResult] = None

# ── Why / Debug trace schemas ────────────────────────────────────────────────

class ConditionTrace(BaseModel):
    type: str
    enabled: bool
    skipped: bool
    actual_value: Optional[float] = None
    operator: Optional[str] = None
    threshold: Optional[float] = None
    passed: Optional[bool] = None

class RuleTrace(BaseModel):
    rule_name: str
    priority: int
    enabled: bool
    action: str
    logic_operator: str
    matched: bool
    conditions: List[ConditionTrace]

class DeviceDebugResult(BaseModel):
    device_name: str
    device_id: int
    switch_entity_id: str
    temp_c: Optional[float] = None
    active_override: Optional[str] = None
    final_decision: SimImmersionResult
    rule_traces: List[RuleTrace]

class LpDecisionTrace(BaseModel):
    mode: str
    reason: str
    status: str
    current_slot_price_pence: Optional[float] = None
    battery_soc: Optional[float] = None
    solar_power_kw: Optional[float] = None
    objective_value: Optional[float] = None
    optimization_time_ms: Optional[float] = None
    last_run: Optional[str] = None

class WhyResponse(BaseModel):
    timestamp: str
    readings: dict
    lp_decision: LpDecisionTrace
    immersion_devices: List[DeviceDebugResult]

File: frontend/src/api/dev.ts
import { api } from './client'

export interface SimPrice {valid_from:string;price_pence:number}
export interface SimImmersionRule {
  rule_name:string;priority:number;action:'ON'|'OFF';logic_operator:'AND'|'OR';is_enabled:boolean
  price_enabled:boolean;price_operator:string;price_threshold_pence:number
  soc_enabled:boolean;soc_operator:string;soc_threshold_percent:number
  solar_enabled:boolean;solar_operator:string;solar_threshold_kw:number
  temp_enabled:boolean;temp_operator:string;temp_threshold_c:number
  time_enabled:boolean;time_start:string;time_end:string
}
export interface SimImmersionInput {
  battery_soc:number;solar_power_kw:number;current_price_pence:number
  current_temp_c:number|null;rules:SimImmersionRule[]
}
export interface SimRequest {
  battery_soc:number;solar_power_kw:number
  live_charge_rate_kw:number|null;live_battery_voltage_v:number|null
  battery_capacity_kwh:number;battery_max_charge_kw:number
  battery_max_discharge_kw:number;battery_efficiency:number
  battery_min_soc:number;battery_max_soc:number;battery_voltage_v:number
  grid_import_limit_kw:number;grid_export_limit_kw:number
  export_price_pence:number;assumed_load_kw:number
  force_charge_threshold_kw:number;force_discharge_threshold_kw:number
  force_discharge_export_min_kw:number;optimization_horizon_hours:number
  prices:SimPrice[];immersion:SimImmersionInput|null
}
export interface SimPeriodResult {
  slot:number;valid_from:string;price_pence:number;solar_kw:number
  charge_kw:number;discharge_kw:number;grid_import_kw:number;grid_export_kw:number
  soc_kwh:number;soc_pct:number
}
export interface SimImmersionResult {action:boolean;source:string;reason:string}
export interface SimResponse {
  recommended_mode:string;decision_reason:string;optimization_status:string
  objective_value:number|null;optimization_time_ms:number
  recommended_discharge_current:number;periods:SimPeriodResult[]
  immersion:SimImmersionResult|null
}

export const DEFAULT_SIM_REQUEST: SimRequest = {
  battery_soc:50,solar_power_kw:3,live_charge_rate_kw:null,live_battery_voltage_v:null,
  battery_capacity_kwh:20,battery_max_charge_kw:10.5,battery_max_discharge_kw:5,
  battery_efficiency:0.95,battery_min_soc:10,battery_max_soc:100,battery_voltage_v:48,
  grid_import_limit_kw:15,grid_export_limit_kw:5,export_price_pence:15,assumed_load_kw:2,
  force_charge_threshold_kw:0.5,force_discharge_threshold_kw:0.5,
  force_discharge_export_min_kw:0.05,optimization_horizon_hours:24,
  prices:[],immersion:null,
}

export function generateFlatPrices(pence:number, n=48):SimPrice[] {
  const base = new Date()
  base.setMinutes(base.getMinutes()<30?0:30,0,0)
  return Array.from({length:n},(_,i)=>{
    const d = new Date(base.getTime()+i*30*60*1000)
    return {valid_from:d.toISOString(),price_pence:pence}
  })
}

export async function runSimulation(req:SimRequest):Promise<SimResponse> {
  return api.post<SimResponse>('/dev/simulate',req)
}
